File: bench/benches/bell.hs
{-# LANGUAGE BangPatterns #-}
import Bench
import qualified Data.Map.Strict as M

-- Bell numbers in base 36 (see bench/benches/bell.py). Fresh memo per rep.
-- checksum = total characters across all rendered lines until one exceeds the
-- limit. Needs bignums -- Integer (native) handles the large intermediates.

base :: Integer
base = 36

-- factorial (Python memoises only for speed; the direct product is exact).
fact :: Integer -> Integer
fact n = product [2 .. n]

choose :: Integer -> Integer -> Integer
choose n k = fact n `div` (fact k * fact (n - k))

-- bell i from a memo holding bell 0 .. i-1.
bellAt :: M.Map Int Integer -> Int -> Integer
bellAt m i
  | i < 2     = 1
  | otherwise = sum [ choose (fromIntegral i - 1) (fromIntegral k) * (m M.! k)
                    | k <- [0 .. i - 1] ]

-- base-36 rendered length of n (>=1); 0 -> 0, matching `while n > 0`.
show36len :: Integer -> Int
show36len = go 0
  where
    go !acc n | n <= 0    = acc
              | otherwise = go (acc + 1) (n `div` base)

-- Drive i upward, growing the memo on demand (fresh per rep), summing the
-- rendered lengths until one exceeds the limit.
bellRun :: Int -> Integer
bellRun limit = go 0 M.empty 0
  where
    go i memo !total =
      let v     = bellAt memo i
          memo' = M.insert i v memo
          len   = show36len v
      in if len > limit
           then total
           else go (i + 1) memo' (total + fromIntegral len)

main :: IO ()
main = bench "bell" (\u -> bellRun (with u 280))

File: bench/benches/bell.java
import java.math.BigInteger;
import java.util.HashMap;

// Bell numbers in base 36 -- a bignum (BigInteger) stress. fresh memo maps per
// rep; checksum = total characters across all rendered lines.
class Main {
    static final String DIGITS = "0123456789abcdefghijklmnopqrstuvwxyz";
    static final BigInteger BASE = BigInteger.valueOf(DIGITS.length());

    static HashMap<Integer, BigInteger> facts;
    static HashMap<Integer, BigInteger> bells;

    static BigInteger fact(int n) {
        BigInteger c = facts.get(n);
        if (c != null) return c;
        BigInteger x = BigInteger.ONE;
        for (int m = n; m > 1; m--) x = x.multiply(BigInteger.valueOf(m));
        facts.put(n, x);
        return x;
    }

    static BigInteger choose(int n, int k) {
        return fact(n).divide(fact(k).multiply(fact(n - k)));
    }

    static BigInteger bell(int n) {
        BigInteger c = bells.get(n);
        if (c != null) return c;
        BigInteger r;
        if (n < 2) r = BigInteger.ONE;
        else {
            r = BigInteger.ZERO;
            for (int k = 0; k < n; k++)
                r = r.add(choose(n - 1, k).multiply(bell(k)));
        }
        bells.put(n, r);
        return r;
    }

    static int show(BigInteger n) {
        int len = 0;
        while (n.signum() > 0) {
            len++;
            n = n.divide(BASE);
        }
        return len;
    }

    static long bellRun(int limit) {
        facts = new HashMap<>();
        bells = new HashMap<>();
        long total = 0;
        for (int i = 0; ; i++) {
            int b = show(bell(i));
            if (b > limit) return total;
            total += b;
        }
    }

    public static void main(String[] a) {
        Bench.bench("bell", () -> bellRun(280));
    }
}
